source/smart_pointer/smart_pointer.hpp

```diff
@@ -23,7 +39,15 @@ class SmartPointer {
    * El recurso también podría ser NULL lo que ocasionaría que el
    * recurso no administre ningún recurso.
    */
-  explicit SmartPointer(Type *resource) : resource_(resource) {}
+  explicit SmartPointer(Type *resource) : resource_(resource) {
+    this->ref_counter = new ReferenceCounter();
+    this->ref_counter->AddRef();
+  }
+
+  SmartPointer() : resource_(nullptr) {
+    ref_counter = new ReferenceCounter();
+    ref_counter->AddRef();
+  }
```

Comment:
Este caso no es utilizado, pero dado que decidiste implementarlo no deberías tener el contador en 1 ya que no está apuntando a ningún recurso.

```suggestion
  SmartPointer() : resource_(nullptr) {
    ref_counter = new ReferenceCounter();
  }
```

Reply:
La razón por la que añado una referencia es porque en mi destructor verifico si ref_counter == 0. Si no añadiera una referencia, nunca entraría a ref_counter == 0 y no destruiría el ReferenceCounter. Además, no tendría sentido que el contador tenga un número negativo.
Supongo que es un caso especial de la forma en que lo he implementado.